platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StartupStateUtils.java

```diff
@@ -0,0 +1,586 @@
+        try {
+            deserializedSignedState = readStateFile(platformContext, savedStateFile.stateFile());
+        } catch (final IOException e) {
+            logger.error(EXCEPTION.getMarker(), "unable to load state file {}", savedStateFile.stateFile(), e);
+
+            final StateConfig stateConfig = platformContext.getConfiguration().getConfigData(StateConfig.class);
+            if (stateConfig.deleteInvalidStateFiles()) {
+                recycleState(recycleBin, savedStateFile);
+                return null;
+            } else {
+                throw new SignedStateLoadingException("unable to load state, this is unrecoverable");
+            }
+        }
+
+        final State state = deserializedSignedState.reservedSignedState().get().getState();
+
+        final Hash oldHash = deserializedSignedState.originalHash();
+        final Hash newHash = rehashTree(state);
+
+        final SoftwareVersion loadedVersion = deserializedSignedState
+                .reservedSignedState()
+                .get()
+                .getState()
+                .getPlatformState()
+                .getPlatformData()
+                .getCreationSoftwareVersion();
+
+        if (oldHash.equals(newHash)) {
+            logger.info(STARTUP.getMarker(), "Loaded state's hash is the same as when it was saved.");
+        } else if (loadedVersion.equals(currentSoftwareVersion)) {
+            logger.error(
+                    EXCEPTION.getMarker(),
+                    "The saved state file {} was created with the current version of the software, "
+                            + "but the state hash has changed. Unless the state was intentionally modified, "
+                            + "this good indicator that there is probably a bug.",
+                    savedStateFile.stateFile());
+        } else {
+            logger.warn(
+                    STARTUP.getMarker(),
+                    "The saved state file {} was created with version {}, which is different than the "
+                            + "current version {}. The hash of the loaded state is different than the hash of the "
+                            + "state when it was first created, which is not abnormal if there have been data "
+                            + "migrations.",
+                    savedStateFile.stateFile(),
+                    loadedVersion,
+                    currentSoftwareVersion);
+        }
+
+        return deserializedSignedState.reservedSignedState();
+    }
+
+    /**
+     * Recycle a state.
+     *
+     * @param recycleBin the recycle bin
+     * @param stateInfo  the state to recycle
+     */
+    private static void recycleState(@NonNull final RecycleBin recycleBin, @NonNull final SavedStateInfo stateInfo) {
+        logger.warn(STARTUP.getMarker(), "Moving state {} to the recycle bin.", stateInfo.stateFile());
+        try {
+            recycleBin.recycle(stateInfo.getDirectory());
+        } catch (final IOException ee) {
```

Comment:
`ee`?

Reply:
This code was originally inline and the variable `e` was already taken. Fixed.